### ISSUE TITLE
[NativeAOT-LLVM] VSD support

### ISSUE THREAD

**Issue by SingleAccretion** [dotnet/runtimelab#2147]
Interface calls in regular NAOT (as well as CoreCLR) are implemented using virtual stub dispatch (the NativeAOT flavor is actually called "cached interface dispatch", shorted as "CID" in code, but their fundamental idea of using per-callsite caches for targets is the same).

In broad strokes, the compiler generates a "cell", struct two pointers in size embedded in the final image, where the first field (generally) contains the target for the dispatch and the second field is (generally) the cell's current state / cache.

Codegen is then given an address of this cell, dereferences the first field, and calls the resulting pointer. It also passed the cell's address as a secret parameter using a non-standard calling convention, which the stub code, interposed between the caller the ultimate callee, uses to update the cell's state, e. g. resolve the target on first dispatch.

For LLVM/WASM the secret parameter presents a bit of a challenge. Ordinary stub interposition code can handle arbitrary callees because it is signature-agnostic, but if we were to generate stubs akin to it with LLVM, we'd need one such stub for each possible signature called through VSD, which would defeat the purpose (of saving on code size) a bit. We could inline the dispatch with something like below:
```scala
void* tgt = RhpResolveTarget(&cell, object);
(*tgt)(args...);
```
Which would then look a lot like `CORINFO_CALL_CODE_POINTER` (but not quite, because of `this`, and we would not want to do this in `getCallInfo` anyway because it would hurt devirtualization). I think this is also similar to what the IL backend does today.

Tasks:
- [x] Verify if we need to need to do anything special w.r.t. atomic updating of cells/cache entries.
  - Nothing to do for Wasm32 as it will support the atomic compare exchange primitive we need (`i64.atomic.rmw.cmpxchg`).
- [x] See what should be done about the very low `OS_PAGE_SIZE` WASM currently has, it will likely significantly pessimize the native allocator used for VSD caches.
  - Page size was upped to 4K.
- [x] Check if we need explicit null checks for VSD calls with LLVM.
  - It depends on how we structure the support. We could throw NRE from the stub.
    - Codegen will do the check for now.

**Comment by SingleAccretion** [comment]
One issue with the inline helper approach is that native code quality is likely going to be pretty bad (since the resolution call will trash the arguments we've just set up). This could possibly be worked around by resolving before setting up the arguments.

**Comment by jkotas** [comment]
Inlining the dispatch like you have suggested makes sense to me.

**Comment by SingleAccretion** [comment]
A version of this was implemented in #2184.
